### PR TITLE
0.1.0 staging > development

### PR DESCRIPTION
**Pull request by mtschoen-unity**
Includes Poly workspace and bugfixes done to support Poly workspace
- GC improvements
- Fix stuck "menu hidden" state in some cases
- Fix auto-hidden main menu invalidating UI hovers
- Basic grouping support beyond prefabs

### CODE REVIEW

File: Tools/TransformTool/TransformTool.cs [hunk 1]
```diff
@@ -436,6 +436,9 @@ public void ProcessInput(ActionMapInput input, ConsumeControlDelegate consumeCon
 
                     if (leftInput.select.wasJustReleased)
                     {
+                        if (rightInput != null && rightInput.select.wasJustReleased)
+                            m_Scaling = false;
```
Comment (mtschoen-unity) [inline Tools/TransformTool/TransformTool.cs]: This was the fix for the "Main Menu Hidden" bug. Apparently it happens pretty frequently that both trigger releases happen in the same frame.

File: Workspaces/PolyWorkspace/PolyWorkspace.cs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+
+            var zoomTooltip = sliderObject.GetComponentInChildren<Tooltip>();
+            if (zoomTooltip)
+                zoomTooltip.tooltipText = "Drag the Handle to Zoom the Asset Grid";
+
+            var scrollHandle = m_PolyUI.scrollHandle;
+            var scrollHandleTransform = scrollHandle.transform;
+            scrollHandleTransform.SetParent(m_WorkspaceUI.topFaceContainer);
+            scrollHandleTransform.localScale = new Vector3(1.03f, 0.02f, 1.02f); // Extra space for scrolling
+            scrollHandleTransform.localPosition = new Vector3(0f, -0.015f, 0f); // Offset from content for collision purposes
```
Comment (mtschoen-unity) [inline Workspaces/PolyWorkspace/PolyWorkspace.cs]: Make consts

File: Workspaces/PolyWorkspace/Scripts/PolyGridAsset.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        Texture2D m_Thumbnail;
+        bool m_Initialized; // Whether the download/import process has started
+        bool m_Importing;
+
+        public PolyAsset asset { get { return m_Asset; } }
+        public GameObject prefab { get { return m_Prefab; } }
+        public Texture2D thumbnail { get { return m_Thumbnail; } }
+        public bool initialized { get { return m_Initialized; } }
+        public long complexity { get { return m_Complexity; } }
+        public override string index { get { return m_Asset.name; } }
```
Comment (mtschoen-unity) [inline Workspaces/PolyWorkspace/Scripts/PolyGridAsset.cs]: Add comment about how name PolyAssset.name is the guid

File: Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+        Action<float> m_SetImportingScale;
+
+        Action<Color> m_SetImportingColor;
+
+        void Awake()
+        {
+            m_CompleteHoverTransition = CompleteHoverTransition;
+            m_SetThumbnailScale = SetThumbnailScale;
+            m_SetImportingScale = SetImportingScale;
+
```
Comment (dunity) [inline Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs]: remove extra line

File: Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+                data.thumbnailImportCompleted -= OnThumbnailImportCompleted;
+            }
+
+            var currentScale = transform.localScale;
+            var targetScale = visible ? m_IconScale * scaleFactor : Vector3.zero;
+
+            while (currentTime < k_TransitionDuration)
+            {
+                currentTime += Time.deltaTime;
+                transform.localScale = Vector3.Lerp(currentScale, targetScale, currentTime / k_TransitionDuration);
```
Comment (dunity) [inline Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs]: This should use SmoothInOutLerpFloat for the lerp amount

File: Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+                    pivotOffset = pivotOffset * scaleFactor + (Vector3.up + Vector3.forward) * 0.5f * k_MaxPreviewScale;
+                }
+            }
+
+            while (currentTime < kDuration - 0.05f)
+            {
+                if (m_DragObject == null)
+                    yield break; // Exit coroutine if m_GrabbedObject is destroyed before the loop is finished
+
+                currentTime = MathUtilsExt.SmoothDamp(currentTime, kDuration, ref currentVelocity, 0.5f, Mathf.Infinity, Time.deltaTime);
```
Comment (dunity) [inline Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs]: Shouldn't this use SmoothInOutLerpFloat instead (so we reach the desired value)

File: Workspaces/ProjectWorkspace/Shaders/StandardDecal.shader [hunk 7]
```diff
@@ -32,11 +32,16 @@
 
 		void surf (Input IN, inout SurfaceOutputStandard o)
 		{
-			fixed4 c = tex2D(_MainTex, IN.uv_MainTex);
-			c *= _Color * c.a;
-			o.Emission = c.rgb;
+            if (IN.uv_MainTex.x > 1 || IN.uv_MainTex.x < 0
```
Comment (dunity) [inline Workspaces/ProjectWorkspace/Shaders/StandardDecal.shader]: Why the odd tabs?

File: Scripts/Interfaces/FunctionalityInjection/IRequestFeedback.cs [hunk 8]
```diff
@@ -1,48 +1,57 @@
...
+        /// </summary>
+        /// <param name="obj">The caller object</param>
+        public static void ClearFeedbackRequests(this IRequestFeedback obj)
+        {
+            clearFeedbackRequests(obj);
+        }
+
+        /// <summary>
+        /// Get a pooled FeedbackRequest object from the system
+        /// </summary>
```
Comment (dunity) [inline Scripts/Interfaces/FunctionalityInjection/IRequestFeedback.cs]: missing proper summary matching other static functions

Comment (mtschoen-unity) [reply Scripts/Interfaces/FunctionalityInjection/IRequestFeedback.cs]: Yeah I forgot how we did param documentation for Actions and just gave up. I'll find something to replicate here.

File: Scripts/Core/EditorVR.Menus.cs [hunk 9]
```diff
@@ -179,6 +179,14 @@ internal void UpdateMenuVisibilities()
                 k_ActiveDeviceData.Clear();
                 Rays.ForEachProxyDevice(deviceData => { k_ActiveDeviceData.Add(deviceData); });
 
+                foreach (var deviceData in k_ActiveDeviceData)
```
[the remark this answers is not in the document]
Comment (mtschoen-unity) [reply Scripts/Core/EditorVR.Menus.cs]: This one's interesting. It is the fix for the "can't interact when menu is temp-hidden" issue. The temporary states were all being cleared right after we evaluate the hidden states, meaning that anything else trying to access the hidden state would only know if it was perma-hidden.

Thus, in MultipleRayInputModule.RaycastSource.isValid, we were not getting proper hide states. Do this clear step right _before_ evaluation and we can check the temporary states at any other time.

File: Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs (hunk 6 above)
Comment (mtschoen-unity) [reply Workspaces/PolyWorkspace/Scripts/PolyGridItem.cs]: Just copied over from AssetGridItem. Will fix.

**Comment by mtschoen-unity** [comment]
Not sure why I can't change the base back to development. Manual merge successful and pushed back to development.

**Pull request closed without merging**